sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinationsHelpers.java

```diff
@@ -167,7 +175,12 @@ public TableDestination getTable(DestinationT destination) {
     @Override
     Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
         throws CannotProvideCoderException {
-      return inner.getDestinationCoderWithDefault(registry);
+      Coder<DestinationT> destinationCoder = getDestinationCoder();
```

Comment:
`DynamicDestinations#getDestinationCoderWithDefault` is commented as:

```
  // Gets the destination coder. If the user does not provide one, try to find one in the coder
  // registry. If no coder can be found, throws CannotProvideCoderException.
```

This code is written with potentially multiple layers of delegation, and I think the correct behavior here is to return the first non-delegated implementation of `getDestinationCoder()` that appears as we move down the delegation chain.

I would argue that the existing behavior is incorrect. Currently, if an implementing class defines a custom return value for `getDestinationCoder`, that value is ignored when you call `getDestinationCoderWithDefault`. My expectation is that `getDestinationCoderWithDefault` would always return the same value as `getDestinationCoder` except in the null case, in which `getDestinationCoderWithDefault` would then attempt to look up a coder in the registry.

So the change here is intended to fix broken behavior. 

It's possible that a user has written a custom class that extends DelegatingDynamicDestinations and relies on the incorrect behavior, but it feels unlikely to me.

For the scope of the coders provided here, I don't believe this change affects behavior (the method was already returning TableDestinationCoderV2 in all cases).

Reply:
It may actually be better to remove this override altogether. `getDestinationCoderWithDefault` is defined as package-private and probably shouldn't be overriden. Instead, we let `getDestinationCoder` handle delegation, and the base implementation of `getDestinationCoderWithDefault` will make the appropriate `getDestinationCoder` call before falling back to the registry.

Reply:
I've removed the override getDestinationCoderWithDefault completely.

Reply:
> I've removed the override getDestinationCoderWithDefault completely.

Looks like that caused some regressions where destination coder couldn't be found from type. I'm looking into it.

Reply:
The override turns out to be necessary because the `DynamicDestinations` instance itself is eventually passed into `extractFromTypeParameters` and the inner class potentially has richer type information than the delegating class.

Reply:
I've put the override method back in place, including the modification of first checking if we have a non-null return from `getDestinationCoder()`.